[Just In Time Messages] Display Just In Time Message on the My Store screen (visual UI only)

joshheald opened the pull request:
<!-- Remember about a good descriptive title. -->

Part of: #7916 
<!-- Id number of the GitHub issue this PR addresses. -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

In order to promote In-Person Payments to relevant merchants, we are adding JITM support in the mobile apps. In the first instance, we will support JITM support on the My Store screen, above the analytics charts.

In this PR, we display the first JITM if any are returned. We se `content.message` as the title, `content.description` as the detail, and `CTA.message` as the button title. The JITM is displayed by reusing the `FeatureAnnouncementCard` UI.

### Interplay with Product Onboarding
There is another project, Product Onboarding, which will use the same space and UI to display a message encouraging merchants to add their first product. See pe5pgL-dX-p2 for full context. This won't be a JITM, and will take precedence over any JITM, at least in the initial implementation.

@rachelmcr has confirmed that this approach looks feasible for the Product Onboarding project.

### Refactoring
Since we're using the same UI in the same place, I've added the JITM UI in a way that it can be easily used by the product onboarding instead. See [`DashboardViewModel.syncAnnouncements()`](https://github.com/woocommerce/woocommerce-ios/blob/30b029af5c0effd3880fce894dc2c1b9277f64b6/WooCommerce/Classes/ViewRelated/Dashboard/DashboardViewModel.swift#L103), adding code to publish, for example, a `ProductOnboardingAnnouncementCardViewModel`, will show the onboarding card.

To achieve this, I've done a partial refactor of the `FeatureAnnouncementCard`, to have it use a new `AnnouncementCardViewModelProtocol` which can be adopted by any view model we need. There is more which could be factored out of this protocol, but that's an exercise for future improvements.

I've also removed the `FeatureAnnouncementCardCell`, as it was tricky to make the changes there, and unnecessary in light of the `HostingTableViewCell<FeatureAnnouncementView>` which I've used in its place. The cell was only used in one place, for the Product Form: after adding a new product, the announcement card is shown to promote setting product upsells and cross-sells.

I've tested that the upsell campaign still works with the new cell. @ealeksandrov – I think you initially implemented the cell I removed here: feel free to take a look over the code as well if you like, but no worries if you've not got the time just now.

## Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->

### JITM on my store

N.B. see pdfdoF-1uc-p2#comment-2581 for details of setting up your store to be eligible for the test JITM

1. With a US store that is eligible for the test JITM, on a debug/alpha build of the app
2. Open the app
3. Observe that the message is shown

Note that dismissal and CTA buttons are not yet functional.

The banner is behind a feature flag: try with a Release build, to make sure that the banner does not show and the network request to `/jetpack/v4/jitm` is not made.

### Existing product upsell/cross-sell campaign still works after removal of cell

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->

### JITM on My Store

![My Store screen with a JITM visible](https://user-images.githubusercontent.com/2472348/198276837-479cfc01-9203-401e-b6a1-1ca31d67a3d8.jpg)

### Product upsell/cross-sell campaign still works after removal of cell

https://user-images.githubusercontent.com/2472348/198277688-02702915-00f7-429c-bc17-9912cee57e01.mp4



---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

joshheald committed:
Expand abbreviations from code review suggestion

joshheald committed:
7916 Add JITM announcement view to dashboard

CTA and dismiss are not implemented in this commit.

joshheald committed:
Use SwiftUI FeatureAnnouncementCard everywhere

joshheald committed:
7916 Use JITM view model with existing UI

`AnnouncementCardViewModelProtocol` leaves it open for the other Product onboarding work to pass its own VM, or a `FeatureAnnouncementCardViewModel`, whichever makes most sense.

There can be further refactoring of this, and it would probably be useful, but I’ll do some in a future PR. In particular, the dismiss handling is a bit specific, and might benefit from being extracted.

joshheald commented:
@rachelmcr Just adding you as an optional reviewer, as mentioned. No changes since I sent this over before. 

The UI appearing is a bit jerky at the moment, I'll look at animating its insertion in a future PR.

ealeksandrov commented:
Thanks for cleaning up `FeatureAnnouncementCardCell`! Unfortunately when testing this PR I still see UI bug that forced me to fallback on UIKit:

https://user-images.githubusercontent.com/3132438/198286294-cbea3b13-4815-4cac-8afd-b082f7bd2f60.mp4

It would be nice to unify all announcement views but I haven't found a fix for this glitch. Happens on iOS 15 and 16, simulator and device.

toupper commented:
@joshheald The code looks great, but I found a couple of UI glitches that need to be fixed in my opinion:

1. When changing orientation to landscape and back to portrait, there's an extra top space above the message:

<img src="https://user-images.githubusercontent.com/1864060/198315215-2106cdb0-97f9-4ee8-8074-606a79c101c3.png" width="375">

2. Each time we navigate to another tab and back to My Store tab the message is added again:

<img src="https://user-images.githubusercontent.com/1864060/198315718-e85714dc-fbcf-4541-99d0-cea6e636fed6.jpeg" width="375">

joshheald committed:
7916 Prevent multiple cards being displayed at once

joshheald committed:
7916 Sync announcements once in view lifecycle

Previously, we were triggering the JITM request every time we switched to the dashboard tab

pachlava commented:
@joshheald 👋 

Apart from issues mentioned by @toupper, I've noticed only one thing:

<img width="324" alt="Screenshot 2022-10-27 at 17 40 58" src="https://user-images.githubusercontent.com/73365754/198323645-eb878e65-e819-42bc-a7bb-0df22d06657a.png">

I think that black background colour makes the message break the design consistency a bit. Possibly, making it grey (as it's done for upsell banner in Orders List) will improve the situation.

rachelmcr commented on WooCommerce/Classes/ViewRelated/Dashboard/DashboardViewController.swift:
```diff
@@ -73,6 +74,13 @@ final class DashboardViewController: UIViewController {
                                               })
     }()
 
+    private var hasAnnouncementFeatureFlag: Bool { ServiceLocator.featureFlagService.isFeatureFlagEnabled(.promptToEnableCodInIppOnboarding)
```
Should this be checking the `.justInTimeMessagesOnDashboard` feature flag?

joshheald replied:
Yes. Thank you! Fixed in e1df523. It escaped my checking of the release build because of the correct flag being used when we make the request.

joshheald committed:
7916 Use listForeground colour to match surrounds

joshheald committed:
7916 Check correct FeatureFlag for showing UI

joshheald commented:
@toupper thank you for the speedy review!

> I found a couple of UI glitches that need to be fixed in my opinion:
> 
> 1. When changing orientation to landscape and back to portrait, there's an extra top space above the message:
> 
> 2. Each time we navigate to another tab and back to My Store tab the message is added again.

## Multiple cards
The second of these was really two bugs: I requested JITMs in `viewWillAppear`, and I didn't properly remove existing cards from display before showing a new one.

These are fixed in c91c2c2 and c3c7d0e respectively.

## Orientation change
This looks like it is iOS 15 only. However, I don't really think it works well in landscape, because it doesn't scroll away with it being in the header.

I think we should hide it in landscape, and make it collapse like the title does, when we scroll in portrait.

joshheald commented:
@pachlava thanks!

> I think that black background colour makes the message break the design consistency a bit. Possibly, making it grey (as it's done for upsell banner in Orders List) will improve the situation.

I agree. Fixed in 85097b4

![dark-mode](https://user-images.githubusercontent.com/2472348/198338264-1896eb0e-2696-4e45-a4e7-3e33efb5f934.jpg)

joshheald committed:
Merge branch 'trunk' into issue/7916-extract-jitm-uiview

# Conflicts:
#	Experiments/Experiments/DefaultFeatureFlagService.swift
#	Experiments/Experiments/FeatureFlag.swift
#	WooCommerce/Classes/ViewRelated/Dashboard/DashboardViewController.swift

joshheald committed:
Revert "Use SwiftUI FeatureAnnouncementCard everywhere"

This reverts commit a9017a820badf5987921f20d23c5bca51c3bb2bf.

joshheald commented:
@ealeksandrov Thanks for checking this, and sorry I didn't spot the scrolling issue!

> Unfortunately when testing this PR I still see UI bug that forced me to fallback on UIKit:
> 
> It would be nice to unify all announcement views but I haven't found a fix for this glitch. Happens on iOS 15 and 16, simulator and device.

It certainly would, what a shame that it doesn't work. I tried a bunch of stuff to fix it too:

- Ruled out the containing stack view being the issue
- Tried `invalidateIntrinsicContentSize()` in the host cell, instead of `layoutIfNeeded()` as per [this article](https://noahgilmore.com/blog/swiftui-self-sizing-cells/)
- Messed with all the constraints on the host cell and the stack in the card View.
- Other stuff that I can't remember now...

No luck though. I've reverted the removal of the UITableViewCell subclass in 7b89fb2, and tested it again, the Linked Products promo works as it did before.

joshheald committed:
7916 Hide announcement card in landscape on phones

The announcement card in the header doesn’t scroll away, which leaves an unusable amount of space in a compact size class for viewing the stats.

Like on the Order List, we can hide the message in landscape.

pachlava reviewed (approved):
Thanks for addressing all the mentioned UI glitches in "My store" tab @joshheald - it looks good now!

I noticed nothing strange after the second round of testing (both iPad and iPhone Simulators), and also confirm that no JITM-related requests are sent when using the Release build config, and they are sent under Debug config (with empty response for non JITM-eligible stores, and expected response for eligible ones). 👍

joshheald committed:
7916 insert announcement card below store name

Previously, the banner could have been added below the error banner instead.

joshheald committed:
7916 Remove unneccesary arrangedSubviewRemoval

joshheald committed:
7916 Remove unneeded iOS 15 padding after rotation

In iOS 15, UIHostingController added vertical padding to the announcement card after rotation to and from landscape.

This PR updates to use a subclass of UIHostingController, which resolves the padding issue in viewDidLayoutSubviews.

joshheald commented:
> ## Orientation change
> This looks like it is iOS 15 only. However, I don't really think it works well in landscape, because it doesn't scroll away with it being in the header.
> 
> I think we should hide it in landscape, and make it collapse like the title does, when we scroll in portrait.

@toupper this is resolved in aa6e6ee

https://user-images.githubusercontent.com/2472348/198359086-9d83e61f-1c97-4653-a33a-2be3001f1f5b.mp4

Please could you re-review?

I'm also going to see if this hosting controller fixes the cell issue...

joshheald commented:
> I'm also going to see if this hosting controller fixes the cell issue...

... no such luck 😞

joshheald committed:
Merge branch 'trunk' into issue/7916-extract-jitm-uiview

joshheald committed:
7916 Hide announcement card when loaded in compact

toupper commented:
The latest commits fixed the issues I mentioned, LGTM! :shipit:
